src/app/EventLogging.h

```diff
@@ -0,0 +1,58 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/ConcreteEventPath.h>
+#include <app/EventLoggingDelegate.h>
+#include <app/EventManagement.h>
+#include <app/data-model/Encode.h>
+#include <app/data-model/List.h> // So we can encode lists
+
+namespace chip {
+namespace app {
+
+template <typename T>
+class EventLogger : public EventLoggingDelegate
+{
+public:
+    EventLogger(const T & aEventData) : mEventData(aEventData){};
+    CHIP_ERROR WriteEvent(chip::TLV::TLVWriter & aWriter) final override
+    {
+        return DataModel::Encode(aWriter, TLV::ContextTag(to_underlying(EventDataIB::Tag::kData)), mEventData);
+    }
+
+private:
+    const T & mEventData;
+};
+
+template <typename T>
+CHIP_ERROR LogEvent(const T & aEventData, EndpointId aEndpoint, EventOptions aEventOptions, EventNumber & aEventNumber)
+{
+    EventLogger<T> eventData(aEventData);
+    ConcreteEventPath path(aEndpoint, aEventData.GetClusterId(), aEventData.GetEventId());
+    // log the actual event
+    aEventNumber              = 0;
```

Comment:
The eventNumber need to be an monotonically increasing unique number

Reply:
Yes, it has been passed in logEvent api in event management,  and set it inside, so it is out Parma, before pass in, I initialize it as 0

Reply:
updated the doxygen on this part, thanks

Reply:
removed this line, thanks